### DISCUSSION TITLE
Renamed MDOCs

### DISCUSSION THREAD

**Discussion by asarnow** [nextpyp/.github#96] [Q&A] [answered]
I collected some data with redundant tilt series names in SerialEM (e.g. ts1.mrc.mdoc), so I made symbolic links renamed with a unique leading string. Most of them are processed successfully but a number of them fail with an error message like:

```
15 | Traceback (most recent call last):
-- | --
16 | File "/opt/pyp/bin/run/pyp", line 3448, in <module>
17 | tomo_swarm(args.path, args.file, args.debug, args.keep, args.skip)
18 | File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
19 | return func(*args, **kwargs)
20 | File "/opt/pyp/bin/run/pyp", line 1446, in tomo_swarm
21 | ] = preprocess.read_tilt_series(
22 | File "/opt/pyp/src/pyp/preprocess/core.py", line 173, in read_tilt_series
23 | raise Exception(f"{tilt_image_filename} indicated inside {name}.mdoc is not found in {project_raw_path}")
24 | Exception: 23dec02b_ts1.mrc indicated inside 23dec02b_ts1.mdoc is not found in /data/veesler/nextpyp/shared/users/asarnow/projects/Spike_all_counting-hIIvI0aF2RZwKfMgOCs6v75z/tomo-preprocessing-f2JIJRcXboknzKLR/raw
```

Are some of my MDOCs being interpreted as MDOC-per-tilt instead of MDOC-per-tilt-series? Might there be an edit I can make to fix that?

**Comment by tofushan** [comment, marked as answer]
Hi, 

If you are using mdoc file to parse metadata (_i.e._, tilt angle, tilt axis angle, acquisition), the tilted images indicated inside the mdoc file need to match your image files. Taking one mdoc file from EMPIAR-10987 as an example, `FE33_3_1837_-51.0_Dec09_23.37.35.mrc` needs to be present so that nextPYP can find the corresponding tilted images. 

In the mdoc file: 
`SubFramePath = X:\Dosefractions\Fabian\2021_12_09_FE33_3\Frames\FE33_3_1837_-51.0_Dec09_23.37.35.mrc`

If you have mdoc file per tilt-series, be sure there are **no other mdoc files having the same leading strings**. For example, `23dec02b_ts1_xxxx.mdoc` will be treated as a mdoc file for a tilted image if you are using `23dec02b_ts1.mdoc` for tilt-series `23dec02b_ts1`. Same rationale applied to mdoc file per tilted image, they need to have the same leading strings.

**Reply by asarnow** [reply]
My MDOC files are all per-series, the MDOC file names are all unique, and all of the tilt images are .tif movies with their own completely unique names.

There are a number of MDOCs named like `23dec02b_ts1.mrc.mdoc` and `23dec05a_ts1.mrc.mdoc` (actually symlinks to `ts1.mrc.mdoc` in two different directories), as well as others that were already unique named like `24feb19a_DA16-2_42.mrc.mdoc` (here 42 is the TS index advanced by SerialEM). The double extensions are something SerialEM does for some reason as you've probably seen before.

I guess the exception happens because [this glob](https://github.com/nextpyp/pyp/blob/2e7a1112391b0751bdca24cfb002369f2b930b4c/src/pyp/preprocess/core.py#L148C47-L148C50) is picking up multiple MDOCs. I haven't been able to work out yet how the basename is being set to end up including multiple MDOCs though.

**Reply by tofushan** [reply]
In your previous message, the error says `Exception: 23dec02b_ts1.mrc indicated inside 23dec02b_ts1.mdoc is not found in /data/veesler/nextpyp/shared/users/asarnow/projects/Spike_all_counting-hIIvI0aF2RZwKfMgOCs6v75z/tomo-preprocessing-f2JIJRcXboknzKLR/raw`. 

Could you check if `23dec02b_ts1.mdoc` contains `23dec02b_ts1.mrc`? You should also see `23dec02b_ts1.mrc` in `/data/veesler/nextpyp/shared/users/asarnow/projects/Spike_all_counting-hIIvI0aF2RZwKfMgOCs6v75z/tomo-preprocessing-f2JIJRcXboknzKLR/raw`.

**Reply by tofushan** [reply]
It would be great if you provide more information. For example, is your data in single tilt-series or movie-frame files? It looks like your mdoc file has the name of single tilt-series, but you provide movie-frame files.

**Reply by asarnow** [reply]
The error message is misleading as that filename is not found within the .mdoc file - in this case it is generated by stripping .mdoc from the .mdoc filename. This happens if multiple .mdoc files are found for the TS name, [here](https://github.com/nextpyp/pyp/blob/2e7a1112391b0751bdca24cfb002369f2b930b4c/src/pyp/preprocess/core.py#L895). 

Sorry if my previous message was not clear, as I said they are all per-series .mdoc files listing a .tif tilt movie for each section.

The only place .mrc occurs in any of my MDOCs is the `ImageFile` field at the top, but those don't contain the prefix (e.g. it's just `ts1.mrc`) and they're really useful for anything outside of SerialEM. (They're binned, unaligned movie sums and potentially reordered - SerialEM displays them during TS collection).

**Reply by tofushan** [reply]
Is your raw data in single tilt-series (`ts1.mrc`) or made up of multiple tilted movies (`ts1_xxx.tif`, `ts1_yyy.tif`....)? If it is later, you may need to add an entry `SubFramePath` into all the sections in the mdoc file, because these entries are used to located the corresponding tilted movies. Note that you need to follow the order for each section when inserting the entry: `TiltAngle` -> `SubFramePath` -> `DateTime`

If the data is in single tilt-series, we just fixed the issue with reading metadata from mdoc files, but it has not yet been updated. You will need to wait until the next release.

**Reply by asarnow** [reply]
> as I said they are all per-series .mdoc files listing a .tif tilt movie for each section.

All the .mdocs have those fields in that order, they are just standard tilt-series .mdocs from SerialEM.

**Reply by tofushan** [reply]
Do you have any other mdoc files that have names like `23dec02b_ts1*.mdoc` and were also picked up by the glob function?

**Reply by asarnow** [reply]
Aha...there are indeed, the 10th through 19th tilt-series, e.g. `23dec02b_ts10.mrc.mdoc`, `23dec02b_ts11.mrc.mdoc`, etc.

The tsX increments are created automatically by SerialEM when you set up tiltseries targets, it will either append a number starting from 1 to the tiltseries root name that is given, or if that name already ends in a number it will continue from there.

**Reply by tofushan** [reply]
Great! We can fix it in the next release. Thanks for reporting!

**Reply by asarnow** [reply]
Awesome, thanks for being patient with me!

**Reply by abartesaghi** [reply]
This was addressed in version [0.6.4](https://nextpyp.app/files/pyp/0.6.4/docs/index.html).
